### PR TITLE
Remove duplicate `Open Here` context menu item upgrading to newer Preview release

### LINKED ISSUE #8290: PowerShell-6.2.0-preview.2-win-x64 creates duplicate "Open here" context menu item (also preview.3)
Steps to reproduce
------------------
Install PowerShell-6.2.0-preview.2-win-x64.msi and select the installation option for "Open here" context menu item in Windows Explorer.

Expected behavior
-----------------
If the "Open here" context menu item option is selected during installation it should add a single item for "Open here" and a single item for "Open here as Administrator" to the "PowerShell 6-Preview" context menu in Windows Explorer.

Conversely, if the "Open here" context menu item option is not selected during installation, it should remove any existing context menu for "PowerShell 6-Preview" from Windows Explorer.

Uninstalling the program should also remove the "PowerShell 6-Preview" context menu from the Registry.

Actual behavior
---------------
The installation creates two "Open here" context menu items. 
Uninstalling the program does not remove the context menu.

image

Environment data
----------------
The following is from PowerShell 6.1.1 because PowerShell 6.2.0-Preview.2 will not launch:
```
PS C:\> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.1.1
PSEdition                      Core
GitCommitId                    6.1.1
OS                             Microsoft Windows 10.0.17763
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
## PR Summary

Explicitly remove the old `Open` regkey that enabled the `Open Here` context menu functionality as it was renamed to `OpenPwsh` and MSI didn't automatically remove the old key.

@bergmeister can you help review as @TravisEz13 is on vacation currently

Fix https://github.com/PowerShell/PowerShell/issues/8290

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Review by bergmeister — APPROVED** [review]
Code Review: Looks ok to me from a code perspective, a manual test for the 2 scenarios (old context menu does exist or not) would not hurt though. Thanks for taking on the effort for this.

About the issue itself:
- I have seen the start menu failing to uninstall in one of the 6.1 previews, which I believe is the root cause to be of the reported issue.
- When I did the PR to tweak the registry setting to fix a subtle bug with it, I manually tested multiple upgrade scenarios by building 2 installers of different minor version and upgrading from one to the other and there was no issue. Therefore I suspect that maybe the simplification of the installation path to just have the simple version number `6` broke that uninstall functionality on upgrade or something similar that impacted the whole installer? It would be interesting to construct a 6.2 installer and upgrade from an actual 6.1 RTM installer to test this scenario more realisticaly and verify this is _actually_ an issue and not an intermittent issue in one of the previews (the nature of previews is simply that one should expect that sometimes something gets left over). This is the only reason why I have a bit of an odd feeling about adding such code but I'm OK with it if that's your preference.
- Given the recent upgrade issue with shared components across RTM and preview, this could've also been a potential culprit. In PR #7701 I fixed for example the long standing issue where the shortcut did not get removed when RTM and preview were installed, which was due to the component GUID being shared and not unique. All of that just shows the potential of where upgrades can go wrong, maybe it would be better to recommend to always uninstall and re-install PowerShell (and potentially even disable upgrade?) if the expense of upgrade problems/maintenance outweighs the simplicity of upgrades? Since `WiX` is being developed by MSFT, maybe an internal code review of the whole installer from experts like `Rob Mensching` (Github handle: `@robmen`) would be a good long term investment.
- What kind of upgrade scenarios will PowerShell support? Just upgrading one minor or multiple patch versions at a time? Or multiple minor versions like e.g. `6.0.5` -> `6.3.1`? If there was a limitation then one could limit the lifetime of such workarounds to avoid accumulating a lot of those things over time.

**Comment by SteveL-MSFT** [comment]
@bergmeister thanks for looking into this.  Our plan is to move to MSIX (post 6.2) so I don't think we want to spend a bunch of time trying getting things perfect now.

File: assets/Product.wxs [hunk 1]
```diff
@@ -217,6 +217,7 @@
               <RegistryKey Root="HKCR" Key="Directory\ContextMenus\$(var.ProductName)$(var.SimpleProductVersion)$(sys.BUILDARCH)\shell\runas\command">
                 <RegistryValue Type="string" Value="[$(var.ProductDirectoryName)]pwsh.exe -NoExit -RemoveWorkingDirectoryTrailingCharacter -WorkingDirectory &quot;%V!&quot; -Command &quot;$host.UI.RawUI.WindowTitle = '$(var.ProductName) $(var.SimpleProductVersion) ($(sys.BUILDARCH))'&quot;"/>
               </RegistryKey>
+              <RemoveRegistryKey Id="RemoveOldOpenKey" Root="HKCR" Key="Directory\ContextMenus\$(var.ProductName)$(var.SimpleProductVersion)$(sys.BUILDARCH)\shell\open" Action="removeOnInstall"/>
```
Comment (iSazonov) [inline assets/Product.wxs]: Are we sure that `$ (var.SimpleProductVersion)` will work after 6.1 -> 6.2?

Comment (bergmeister) [reply assets/Product.wxs]: I thought `$(var.SimpleProductVersion)` is just `6` now?

Comment (iSazonov) [reply assets/Product.wxs]: Thanks! I see that in Build.psm1.

**Review by iSazonov — APPROVED** [review]

**Comment by SteveL-MSFT** [comment]
@iSazonov I think we're good to merge this one

**Comment by Karl-WE** [comment]
This issue will still affect users that upgrade to the latest PoSh Core 6.2.x and once had a preview release.
It will also not be cleared for those that upgrade any PoSh Core Preview.

**Suggestion** 
please implement a cleanup with the next release of PoSH 6.x or 7.x, no matter if the user checks the explorer integration 
The MSI can check if preview-6 is still installed by wmi check and directory check on
C:\Program Files\PowerShell\6-preview

if the directory and WMI call results there is no 6 preview installed cleanup should purge the orphaned registry keys.

image

**Comment by iSazonov** [comment]
@Karl-WE If I remember correctly a conclusion was  to do not fix the old and rare issue. Also taking in account how msi works it is not easy.

**Comment by Karl-WE** [comment]
I have searched the entire registry for 
"PowerShell 6-preview" but no finds. 

How can we remove this manually?

**Comment by iSazonov** [comment]
Yes, of cause. You need to search "pwsh.exe"

**Comment by bergmeister** [comment]
The registry keys are here:
https://github.com/PowerShell/PowerShell/blob/bf5956193fde80da118446bf92fede0035cf5014/assets/Product.wxs#L186
Alternatively, you can also just install a preview of 6 again with the explorer context menu again and uninstall it explicitly, which will always remove the registry entries, it is just the case of using an upgrade install, which can leave registry entries behind. I agree with @isazonov that we should not need to invest effort for more polished previews, that's just the nature of previews (saving time for the dev team to get early feedback without having to invest time in proper support).

**Comment by Karl-WE** [comment]
`<# 
.SYNOPSIS 
Deletes orphaned Explorer, Drive and Context Menu entries of former Powershell core installations

.NOTES 
This script was made spending too much of time but learning some more PoSh
instead of simply install and uninstall Powershell Core 6 Preview as suggested

.AUTHOR 
Karl Wester-Ebbinghaus (@tweet_alqamar)

.CREDITS
https://github.com/bergmeister for linking registry keys from the code
https://social.msdn.microsoft.com/profile/Benjamin+Armstrong for the always useful self-evelation check and execution

.SOURCES
https://github.com/PowerShell/PowerShell/blob/bf5956193fde80da118446bf92fede0035cf5014/assets/Product.wxs#L186
https://blogs.msdn.microsoft.com/virtual_pc_guy/2010/09/23/a-self-elevating-powershell-script/

.VERSION
1.0 

.VERSIONHISTORY
1.0 - initial version

.TODO
- implement parameters
- use a function
- upload to github

#>


# Get the ID and security principal of the current user account
$myWindowsID=[System.Security.Principal.WindowsIdentity]::GetCurrent()
$myWindowsPrincipal=new-object System.Security.Principal.WindowsPrincipal($myWindowsID)
 
# Get the security principal for the Administrator role
$adminRole=[System.Security.Principal.WindowsBuiltInRole]::Administrator
 
# Check to see if we are currently running "as Administrator"
if ($myWindowsPrincipal.IsInRole($adminRole))
   {
   # We are running "as Administrator" - so change the title and background color to indicate this
   $Host.UI.RawUI.WindowTitle = $myInvocation.MyCommand.Definition + "(Elevated)"
   $Host.UI.RawUI.BackgroundColor = "DarkBlue"
   clear-host
   }
else
   {
   # We are not running "as Administrator" - so relaunch as administrator
   
   # Create a new process object that starts PowerShell
   $newProcess = new-object System.Diagnostics.ProcessStartInfo "PowerShell";

   # Specify the current script path and name as a parameter

 #$newProcess.Arguments = "& '" + $myInvocation.MyCommand.Definition + "'";

 $newProcess.Arguments = "& '" + $script:MyInvocation.MyCommand.Path + "'"

 # Indicate that the process should be elevated

 $newProcess.Verb = "runas";

 # Start the new process

 [System.Diagnostics.Process]::Start($newProcess);

   # Exit from the current, unelevated, process
   exit
   }

# Run your code that needs to be elevated here

#variables
[array]$regroot = @("HKLM:",
                    "HKCU:")

[array]$regpath = @("\SOFTWARE\Classes\Directory\background\shell\",
                    "\SOFTWARE\Classes\Directory\ContextMenus\",
                    "\SOFTWARE\Classes\Directory\shell\",
                    "\SOFTWARE\Classes\Drive\shell\")


[string]$ProductName = "Powershell"
[int]$SimpleProductVersion = "6"
[string]$branch = "-preview"
[string]$BUILDARCH = "x64"
[string]$buildstring = $ProductName + $SimpleProductVersion + $branch + $BUILDARCH


foreach ($root in $regroot) {
    foreach ($path in $regpath){
        
     #removing obsolete items from registry
     remove-item -path ($root + $path + $buildstring) -Confirm -Recurse -ErrorAction SilentlyContinue 

     #debug creating items
     #new-item -path ($root + $path + $buildstring) -Confirm -ErrorAction SilentlyContinue 
    }
}`

**Comment by sba923** [comment]
I wrote something similar that will remove entries for all PowerShell versions that are no longer installed. Hope this helps.

[Remove-PowerShellOpenWithEntries.zip](https://github.com/PowerShell/PowerShell/files/4643554/Remove-PowerShellOpenWithEntries.zip)

**Comment by o-l-a-v** [comment]
I made a little PowerShell script to clean up right click context menus as well.
Run with ```-WriteChanges``` to actually remove anything.

```powershell
#Requires -RunAsAdministrator
<#
    .SYNOPSIS
        Clean up right click context menu.

    .NOTES
        Author:   Olav Rønnestad Birkeland
        Created:  210828
        Modified: 210828

    .EXAMPLE
        # Intel Graphics Settings
        & $psISE.'CurrentFile'.'FullPath' -NamesLike 'igfx*'
        & $psISE.'CurrentFile'.'FullPath' -NamesLike 'igfx*' -WriteChanges
        # PowerShell 7 Preview
        & $psISE.'CurrentFile'.'FullPath' -NamesLike 'PowerShell7-preview*'
        & $psISE.'CurrentFile'.'FullPath' -NamesLike 'PowerShell7-preview*' -WriteChanges      
#>


# Input parameters
[OutputType($null)]
Param(
    [Parameter(Mandatory, HelpMessage = 'String array of PowerShell -like statements to match with.')]
    [string[]] $NamesLike,

    [Parameter(HelpMessage = 'Whether to write changes.')]
    [switch] $WriteChanges
)


# PowerShell preferences
$ErrorActionPreference = 'Stop'
$InformationPreference = 'Continue'


# Assets
$Paths = 


# Loop through
$(
    [string[]](
        'HKEY_CLASSES_ROOT',
        'HKEY_LOCAL_MACHINE\SOFTWARE\Classes'
    )
).ForEach{
    [string[]](
        ('Registry::{0}\Directory\background\shell' -f $_),
        ('Registry::{0}\Directory\Background\shellex\ContextMenuHandlers' -f $_),
        ('Registry::{0}\Directory\ContextMenus' -f $_),
        ('Registry::{0}\Directory\shell' -f $_),
        ('Registry::{0}\Drive\shell' -f $_)
    )
}.ForEach{
    Get-ChildItem -Path $_ -Depth 0
}.'PSPath'.Where{
    $(
        foreach ($NameLike in $NamesLike) {
            $_.Split('\')[-1] -like $NameLike
        }
    ) -contains $true
}.ForEach{
    Write-Information -MessageData $_
    if ($WriteChanges) {
        if (Test-Path -Path $_) {
            $null = Remove-Item -Path $_ -Recurse -Force
            Write-Information -MessageData ('{0}Successfully removed.' -f "`t")
            $null = Start-Sleep -Milliseconds 100
        }
        else {
            Write-Information -MessageData ('{0}Already removed.' -f "`t")
        }
    }
}
```

**Comment by Karl-WE** [comment]
Hi Steve @SteveL-MSFT do you think one these scripts could be included in the MSI? Not sure about Win11 as so far the integration due Windows Terminal is better and context menu only includes Terminal by default.

**Pull request merged**